File: website/Services/FirebaseService.cs
using Firebase.Database;
using Firebase.Database.Query;
using website.Data;
using System.Linq;
using System.Threading.Tasks;
using System.Collections.Generic;
using System;

namespace website.Services
{
    public class FirebaseService
    {
        private const string FirebaseDatabaseUrl = "https://sharecircle-f335e-default-rtdb.asia-southeast1.firebasedatabase.app/";
        private readonly FirebaseClient _firebaseClient;

        // In-memory cache
        private List<User> _cachedUsers;
        private DateTime _lastUsersFetchTime;
        private List<DonationPost> _cachedPosts;
        private DateTime _lastPostsFetchTime;
        private const int CacheDurationMinutes = 5;

        public FirebaseService()
        {
            _firebaseClient = new FirebaseClient(FirebaseDatabaseUrl);
        }

        public async Task<List<User>> GetAllUsersAsync()
        {
            try
            {
                var users = await _firebaseClient
                    .Child("users")
                    .OnceAsync<User>();

                return users.Select(u => 
                {
                    var user = u.Object;
                    user.UserId = u.Key; 
                    return user;
                }).ToList();
            }
            catch
            {
                return new List<User>();
            }
        }

        public async Task<List<User>> GetRecentUsersAsync(int count, bool forceRefresh = false)
        {
            if (!forceRefresh && _cachedUsers != null && DateTime.Now < _lastUsersFetchTime.AddMinutes(CacheDurationMinutes))
            {
                return _cachedUsers;
            }

            try
            {
                List<User> result = new List<User>();

                // Strategy 1: Try to order by CreatedAt (requires index)
                try 
                {
                    var users = await _firebaseClient
                        .Child("users")
                        .OrderBy("CreatedAt")
                        .LimitToLast(count)
                        .OnceAsync<User>();

                    var list = users.Select(u => 
                    {
                        var user = u.Object;
                        user.UserId = u.Key; 
                        return user;
                    }).ToList();

                    list.Reverse(); // Newest first
                    result = list;
                }
                catch (Exception ex)
                {
                    Console.WriteLine($"[GetRecentUsersAsync] Strategy 1 failed: {ex.Message}");
                    
                    // Strategy 2: Order by Key (default index)
                    try
                    {
                        var users = await _firebaseClient
                            .Child("users")
                            .OrderByKey()
                            .LimitToLast(count)
                            .OnceAsync<User>();

                        var list = users.Select(u => 
                        {
                            var user = u.Object;
                            user.UserId = u.Key; 
                            return user;
                        }).ToList();

                        list.Reverse();
                        result = list;
                    }
                    catch (Exception ex2)
                    {
                        Console.WriteLine($"[GetRecentUsersAsync] Strategy 2 failed: {ex2.Message}");

                        // Strategy 3: Fallback to fetching all (slow but reliable)
                        Console.WriteLine("[GetRecentUsersAsync] Falling back to GetAllUsersAsync");
                        var allUsers = await GetAllUsersAsync();
                        result = allUsers.TakeLast(count).Reverse().ToList();
                    }
                }

                // Update cache
                if (result.Any()) 
                {
                    _cachedUsers = result;
                    _lastUsersFetchTime = DateTime.Now;
                }
                
                return result;
            }
            catch (Exception ex)
            {
                Console.WriteLine($"[GetRecentUsersAsync] All strategies failed: {ex.Message}");
                return new List<User>();
            }
        }

        public async Task<List<DonationPost>> GetAllPostsAsync()
        {
            try
            {
                var posts = await _firebaseClient
                    .Child("donations")
                    .OnceAsync<DonationPost>();

                return posts.Select(p => 
                {
                    var post = p.Object;
                    post.PostId = p.Key;
                    return post;
                }).ToList();
            }
            catch
            {
                return new List<DonationPost>();
            }
        }

        public async Task<List<DonationPost>> GetRecentPostsAsync(int count, bool forceRefresh = false)
        {
            if (!forceRefresh && _cachedPosts != null && DateTime.Now < _lastPostsFetchTime.AddMinutes(CacheDurationMinutes))
            {
                return _cachedPosts;
            }

            try
            {
                List<DonationPost> result = new List<DonationPost>();

                // Strategy 1: Try to order by CreatedAt
                try 
                {
                    var posts = await _firebaseClient
                        .Child("donations")
                        .OrderBy("CreatedAt")
                        .LimitToLast(count)
                        .OnceAsync<DonationPost>();

                    var list = posts.Select(p => 
                    {
                        var post = p.Object;
                        post.PostId = p.Key;
                        return post;
                    }).ToList();

                    list.Reverse();
                    result = list;
                }
                catch (Exception ex)
                {
                    Console.WriteLine($"[GetRecentPostsAsync] Strategy 1 failed: {ex.Message}");
                    
                    // Strategy 2: Order by Key
                    try
                    {
                        var posts = await _firebaseClient
                            .Child("donations")
                            .OrderByKey()
                            .LimitToLast(count)
                            .OnceAsync<DonationPost>();

                        var list = posts.Select(p => 
                        {
                            var post = p.Object;
                            post.PostId = p.Key;
                            return post;
                        }).ToList();

                        list.Reverse();
                        result = list;
                    }
                    catch (Exception ex2)
                    {
                        Console.WriteLine($"[GetRecentPostsAsync] Strategy 2 failed: {ex2.Message}");

                        // Strategy 3: Fallback to fetching all
                        Console.WriteLine("[GetRecentPostsAsync] Falling back to GetAllPostsAsync");
                        var allPosts = await GetAllPostsAsync();
                        result = allPosts.TakeLast(count).Reverse().ToList();
                    }
                }

                // Update cache
                if (result.Any())
                {
                    _cachedPosts = result;
                    _lastPostsFetchTime = DateTime.Now;
                }

                return result;
            }
            catch (Exception ex)
            {
                Console.WriteLine($"[GetRecentPostsAsync] All strategies failed: {ex.Message}");
                return new List<DonationPost>();
            }
        }

        public async Task<bool> DeleteUserAsync(string userId)
        {
            try
            {
                // Note: Deleting a user from DB doesn't delete from Auth, 
                // but for this admin panel we mainly manage DB records.
                await _firebaseClient.Child("users").Child(userId).DeleteAsync();
                
                // Remove from cache if exists
                if (_cachedUsers != null)
                {
                    var item = _cachedUsers.FirstOrDefault(u => u.UserId == userId);
                    if (item != null)
                    {
                        _cachedUsers.Remove(item);
                    }
                }
                
                return true;
            }
            catch
            {
                return false;
            }
        }

        public async Task<bool> DeletePostAsync(string postId)
        {
            try
            {
                await _firebaseClient.Child("donations").Child(postId).DeleteAsync();

                // Remove from cache if exists
                if (_cachedPosts != null)
                {
                    var item = _cachedPosts.FirstOrDefault(p => p.PostId == postId);
                    if (item != null)
                    {
                        _cachedPosts.Remove(item);
                    }
                }

                return true;
            }
            catch
            {
                return false;
            }
        }
        public async Task<bool> AuthenticateAdminAsync(string username, string password)
        {
            // Fallback for hardcoded admins
            if ((username == "admin" && password == "admin123") || 
                (username == "nwchang" && password == "admin1234"))
            {
                return true;
            }

            try
            {
                var adminNode = await _firebaseClient
                    .Child("admins")
                    .Child(username)
                    .OnceSingleAsync<dynamic>(); // Using dynamic to avoid creating a new class

                if (adminNode != null)
                {
                    // Check if password matches
                    string dbPassword = adminNode.password;
                    return dbPassword == password;
                }
                
                return false;
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Auth Error: {ex.Message}");
                return false;
            }
        }
    }
}

File: website-nodejs/services/firebaseService.js
// Firebase REST API Service (无需服务账号)
const https = require('https');

const FIREBASE_DATABASE_URL = 'https://sharecircle-f335e-default-rtdb.asia-southeast1.firebasedatabase.app';

// Cache settings
let cachedUsers = null;
let lastUsersFetchTime = null;
let cachedPosts = null;
let lastPostsFetchTime = null;
const CACHE_DURATION_MINUTES = 5;

// Helper function to make Firebase REST API requests
function firebaseRequest(path, method = 'GET', data = null) {
    return new Promise((resolve, reject) => {
        const url = new URL(`${FIREBASE_DATABASE_URL}${path}.json`);

        const options = {
            hostname: url.hostname,
            path: url.pathname + url.search,
            method: method,
            headers: {
                'Content-Type': 'application/json'
            }
        };

        const req = https.request(options, (res) => {
            let body = '';
            res.on('data', chunk => body += chunk);
            res.on('end', () => {
                try {
                    const result = JSON.parse(body);
                    resolve(result);
                } catch (e) {
                    resolve(null);
                }
            });
        });

        req.on('error', (e) => {
            console.error('Firebase request error:', e);
            reject(e);
        });

        req.setTimeout(10000, () => {
            req.destroy();
            reject(new Error('Request timeout'));
        });

        if (data) {
            req.write(JSON.stringify(data));
        }
        req.end();
    });
}

class FirebaseService {
    // Get all users
    async getAllUsers() {
        try {
            const data = await firebaseRequest('/users');
            if (!data) return [];

            return Object.entries(data).map(([key, value]) => ({
                userId: key,
                ...value
            }));
        } catch (error) {
            console.error('Error fetching users:', error);
            return [];
        }
    }

    // Get recent users with caching
    async getRecentUsers(count = 50, forceRefresh = false) {
        const now = new Date();
        if (!forceRefresh && cachedUsers && lastUsersFetchTime &&
            (now - lastUsersFetchTime) < CACHE_DURATION_MINUTES * 60 * 1000) {
            return cachedUsers;
        }

        try {
            const users = await this.getAllUsers();
            const result = users.slice(-count).reverse();

            cachedUsers = result;
            lastUsersFetchTime = now;
            return result;
        } catch (error) {
            console.error('Error fetching recent users:', error);
            return cachedUsers || [];
        }
    }

    // Get all posts
    async getAllPosts() {
        try {
            const data = await firebaseRequest('/donations');
            if (!data) return [];

            return Object.entries(data).map(([key, value]) => ({
                postId: key,
                ...value
            }));
        } catch (error) {
            console.error('Error fetching posts:', error);
            return [];
        }
    }

    // Get recent posts with caching
    async getRecentPosts(count = 50, forceRefresh = false) {
        const now = new Date();
        if (!forceRefresh && cachedPosts && lastPostsFetchTime &&
            (now - lastPostsFetchTime) < CACHE_DURATION_MINUTES * 60 * 1000) {
            return cachedPosts;
        }

        try {
            const posts = await this.getAllPosts();
            const result = posts.slice(-count).reverse();

            cachedPosts = result;
            lastPostsFetchTime = now;
            return result;
        } catch (error) {
            console.error('Error fetching recent posts:', error);
            return cachedPosts || [];
        }
    }

    // Delete user
    async deleteUser(userId) {
        try {
            await firebaseRequest(`/users/${userId}`, 'DELETE');

            // Update cache
            if (cachedUsers) {
                cachedUsers = cachedUsers.filter(u => u.userId !== userId);
            }
            return true;
        } catch (error) {
            console.error('Error deleting user:', error);
            return false;
        }
    }

    // Delete post
    async deletePost(postId) {
        try {
            await firebaseRequest(`/donations/${postId}`, 'DELETE');

            // Update cache
            if (cachedPosts) {
                cachedPosts = cachedPosts.filter(p => p.postId !== postId);
            }
            return true;
        } catch (error) {
            console.error('Error deleting post:', error);
            return false;
        }
    }

    // Authenticate admin
    async authenticateAdmin(username, password) {
        try {
            const adminData = await firebaseRequest(`/admins/${username}`);
            console.log(`[Auth] Check for ${username}:`, adminData);

            if (adminData) {
                // Handle potential data entry error in DB where key is "password:"
                const dbPassword = adminData.password || adminData['password:'];
                if (dbPassword === password) {
                    return true;
                }
            }
            return false;
        } catch (error) {
            console.error('Auth error:', error);
            return false;
        }
    }
}

module.exports = new FirebaseService();
